staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -106,9 +110,17 @@ func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
 		// See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
 		_, err := client.Get(ctx, path.Join("/", c.Prefix, "health"))
 		if err == nil {
-			return nil
+			healthzErrorMessage.Store("")
+			return
+		}
+		healthzErrorMessage.Store(err.Error())
+	}, time.Second, wait.NeverStop)
```

Comment:
if /healthz was hit less than once per second previously, this poll increases reads...

Reply:
> if /healthz was hit less than once per second previously, this poll increases reads...

To a degree we're concerned about (I'm not really)?  I'm open to other intervals.  I would rather have a constant time /heathz return

Reply:
does this create a new connection every time or reuses the existing one?

Reply:
> does this create a new connection every time or reuses the existing one?

Not sure what etcd client does under the covers.  Naively I would expect re-use.  This does what the healthz endpoint did previously.

Reply:
I'm a big +1 on constant time /healthz return.

For the check frequency, my first thought was that we should try to hit etcd at a rate related to the /healthz frequency, with some rate limit (e.g. poll at 1/n the rolling average of healthz calls over the last m minutes, capped at one second). Ideally, if kube-apiserver was being checked once a minute, we'd check etcd at some subdivision of that, and if /healthz was never called, this wouldn't hit etcd at all.

I don't feel super strongly, though, and maybe I'm overthinking it... cc @lavalamp who might have more specific thoughts about the behavior of this check in a server at rest in steady state.

Reply:
> > does this create a new connection every time or reuses the existing one?
> 
> Not sure what etcd client does under the covers. Naively I would expect re-use. This does what the healthz/etcd endpoint did previously.

You have to look at line 86. We're creating a new client each time so I would actually not expect connection reuse. (we should probably fix that though)

Reply:
Ah never mind it stores the value (@liggitt pointed that out), I hadn't read that far.

Reply:
I regret never stopping things every time I add them, can we make this shut down able? I expect this to add a lot of useless goroutines to the integration tests, but even if I couldn't think of a reason I just always, always regret that.

Reply:
(Sorry, that comment is about the line, not a response to this conversation)

Reply:
re: frequency, 1/s is probably extreme, I'd go like 2-3x the speed of whatever we configure the default readiness probe at? IIRC that's ~15s, so maybe 5-7s between checks would be good.

Reply:
> re: frequency, 1/s is probably extreme, I'd go like 2-3x the speed of whatever we configure the default readiness probe at? IIRC that's ~15s, so maybe 5-7s between checks would be good.

I can reduce to 5s.

Reply:
I've implemented this approach here https://github.com/kubernetes/kubernetes/pull/112046